plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/launcher/PipelineArgumentsTab.java

```diff
@@ -298,50 +312,97 @@ PipelineRunner getSelectedRunner() {
+      clearRunners();
+      defaultOptionsComponent.setEnabled(false);
+      userOptionsSelector.setEnabled(false);
+      pipelineOptionsForm.updateForm(null, null);
+      return;
+    }
 
-      updateRunnerButtons(majorVersion);
-      updateHierarchy(majorVersion);
+    updateRunnerButtons(launchConfiguration);
 
-      defaultOptionsComponent.setUseDefaultValues(launchConfiguration.isUseDefaultLaunchOptions());
-      defaultOptionsComponent.setPreferences(getPreferences());
-      defaultOptionsComponent.setCustomValues(launchConfiguration.getArgumentValues());
+    defaultOptionsComponent.setEnabled(true);
+    defaultOptionsComponent.setUseDefaultValues(launchConfiguration.isUseDefaultLaunchOptions());
+    defaultOptionsComponent.setPreferences(getPreferences());
+    defaultOptionsComponent.setCustomValues(launchConfiguration.getArgumentValues());
 
-      String userOptionsName = launchConfiguration.getUserOptionsName();
-      userOptionsSelector.setText(Strings.nullToEmpty(userOptionsName));
+    userOptionsSelector.setEnabled(true);
+    String userOptionsName = launchConfiguration.getUserOptionsName();
+    userOptionsSelector.setText(Strings.nullToEmpty(userOptionsName));
 
-      updatePipelineOptionsForm();
+    updatePipelineOptionsForm();
+  }
+
+  /**
+   * Reload any computed information only if the launch configuration has changed in some meaningful
+   * way. This must be a fast check as this method is called from
+   * {{@link #isValid(ILaunchConfiguration)}}, which is called frequently.
+   * 
+   * @return true if values were reloaded, or false if the configuration was up-to-date
+   */
+  @VisibleForTesting
+  boolean reload(ILaunchConfiguration configuration) {
+    try {
+      // recompute the features of interest from the provided launch configuration
+      IProject project = findProject(configuration);
+      MajorVersion majorVersion = project == null || !project.isAccessible() ? null
+          : dependencyManager.getProjectMajorVersion(project);
+      PipelineLaunchConfiguration launchConfiguration = majorVersion == null ? null
+          : PipelineLaunchConfiguration.fromLaunchConfiguration(majorVersion, configuration);
+      if (Objects.equals(project, this.project)
+          && Objects.equals(launchConfiguration, this.launchConfiguration)) {
+        // our features of interest are the same
+        return false;
+      }
+      this.project = project;
+      this.launchConfiguration = launchConfiguration;
+      updateHierarchy();
+      return true;
     } catch (CoreException | InvocationTargetException | InterruptedException ex) {
       // TODO: Handle
-      DataflowUiPlugin.logError(ex, 
-          "Error while initializing from existing configuration"); //$NON-NLS-1$
+      DataflowUiPlugin.logError(ex, "Error while initializing from existing configuration"); //$NON-NLS-1$
+      project = null;
+      launchConfiguration = null;
+      return true;
```

Comment:
Maybe `return false`.

Reply:
Returning false means the values were not changed.  Added a comment.